baseten-performance-client-typescript/baseten-performance-client/src/lib.rs

```diff
@@ -0,0 +1,1328 @@
+      .decode_utf8()
+      .map_err(|_| value_error("Invalid UTF-8 in URL path"))?;
+
+    // Check both encoded and decoded forms for traversal patterns
+    let patterns = [
+      "..", "./", "/.", "%2e%2e", "%2E%2E", "%2e.", ".%2e", "%2f", "%5c",
+    ];
+    let decoded_str = decoded.as_ref();
+
+    for pattern in &patterns {
+      if path.contains(pattern) || decoded_str.contains(pattern) {
+        return Err(value_error("URL path contains invalid traversal patterns"));
+      }
+    }
+
+    // Check for null bytes and control characters
+    if decoded_str
+      .chars()
+      .any(|c| c == '\0' || (c.is_control() && c != '\t'))
+    {
+      return Err(value_error("URL path contains invalid control characters"));
+    }
+
+    // Ensure path doesn't start with multiple slashes (//)
+    if path.starts_with("//") {
+      return Err(value_error("URL path cannot start with double slashes"));
+    }
+
+    Ok(())
+  }
+  fn get_api_key(api_key: Option<String>) -> napi::Result<String> {
+    api_key
+      .or_else(|| std::env::var("BASETEN_API_KEY").ok())
+      .or_else(|| std::env::var("OPENAI_API_KEY").ok())
+      .ok_or_else(|| value_error("API key not provided and no BASETEN/OPENAI env variable found"))
+  }
+
+  fn validate_timeout(timeout_s: f64) -> napi::Result<Duration> {
+    if !(MIN_REQUEST_TIMEOUT_S..=MAX_REQUEST_TIMEOUT_S).contains(&timeout_s) {
+      return Err(value_error(format!(
+        "Timeout {:.3}s is outside the allowed range [{:.3}s, {:.3}s]",
+        timeout_s, MIN_REQUEST_TIMEOUT_S, MAX_REQUEST_TIMEOUT_S
+      )));
+    }
+    Ok(Duration::from_secs_f64(timeout_s))
+  }
+
+  fn validate_concurrency(max_concurrent_requests: usize, batch_size: usize) -> napi::Result<()> {
+    if max_concurrent_requests == 0 || max_concurrent_requests > MAX_CONCURRENCY_HIGH_BATCH {
+      return Err(value_error(format!(
+        "max_concurrent_requests must be >0 and <= {}",
+        MAX_CONCURRENCY_HIGH_BATCH
+      )));
+    }
+    if batch_size == 0 || batch_size > MAX_BATCH_SIZE {
+      return Err(value_error(format!(
+        "batch_size must be >0 and <= {}",
+        MAX_BATCH_SIZE
+      )));
+    }
+    if max_concurrent_requests > MAX_CONCURRENCY_LOW_BATCH
+      && batch_size < CONCURRENCY_HIGH_BATCH_SWITCH
+    {
+      return Err(value_error(format!(
+        "max_concurrent_requests must be < {} when batch_size < {} (be nice to the server)",
+        MAX_CONCURRENCY_LOW_BATCH, CONCURRENCY_HIGH_BATCH_SWITCH
+      )));
+    }
+    Ok(())
+  }
+}
+
+#[napi]
+impl PerformanceClient {
+  #[napi(constructor)]
+  pub fn new(base_url: String, api_key: Option<String>) -> napi::Result<Self> {
+    // Comprehensive URL validation and normalization
+    let normalized_base_url = Self::validate_and_normalize_url(&base_url)?;
+
+    let api_key_str = Self::get_api_key(api_key)?;
+
+    // Configure high-performance HTTP client
+    let client = Client::builder()
```

Comment:
in the latest main, we noticed there is a regression where http2 is slower due to the way it does multiplexing. The solution to it was releases last week. 

I think this PR is fine as is, and I am thinking of moving baseten-performance-client-typescript + baseten-performance-client together to one path.